### PR TITLE
messages from roothash mux app

### LINKED ISSUE #2230: Registration panics if we stop the compute worker early
like if we exit really quickly. which, when we do, it closes the compute worker node's initCh without setting a workerHost

## Details

and then the registration callback tries to access the workerHost and segfaults.

## Acceptance Criteria
* that we handle this case

### LINKED ISSUE #2231: broadcastTx panics during shutdown
the when we try to create a subscription, we get a nil (internally, but as I learned, not `== nil`) Subscription if tendermint is shutting down

## Details

and then later we panic on `txSub.Out()`

## Acceptance Criteria
* that we handle this case

### PR DESCRIPTION
**Pull request by pro-wh**
This adds a "roothash messages" field to the compute pipeline. A runtime would be able to send them, and they propagate into a block header.

Roothash messages are stored inline in the block header, rather than in storage.

This implements the simple merging algorithm, where up to one compute committee can send roothash messages.

This adds the staking general balance adjustment message, which causes the staking mux app to adjust an account's general balance.

There's a dummy acceptable transfer peer policy to control what runtimes can send these staking general balance adjustment messages. It accesses a set of approved runtime IDs in the genesis document.

TODO:
- [x] switch dummy acceptable transfer peer policy to deny all
- [x] remove  dummy roothash message types
- [x] test unacceptable transfer peer
- [x] come up with a way to test without hacking up successfulRound

blocks https://github.com/oasislabs/ekiden/issues/2077
context https://github.com/oasislabs/private-rfcs/pull/144

fixes #2230
fixes #2231

### CODE REVIEW

**Comment by pro-wh** [comment]
added an e2e test. please review

File: go/oasis-node/cmd/debug/byzantine/compute.go [hunk 1]
```diff
@@ -161,6 +162,8 @@ func (cbc *computeBatchContext) createCommitment(id *identity.Identity, rak sign
 		PreviousHash: cbc.bd.Header.EncodedHash(),
 		IORoot:       cbc.newIORoot,
 		StateRoot:    cbc.newStateRoot,
+		// TODO: allow script to set roothash messages?
```
Comment (kostko) [inline go/oasis-node/cmd/debug/byzantine/compute.go]: File issue for this?

File: go/oasis-test-runner/scenario/e2e/roothash_messages.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+		return errors.Wrap(err, "setting epoch")
+	}
+	logger.Info("epoch transition done")
+
+	select {
+	case err = <-sc.net.Errors():
+		_ = client.Process.Kill()
+		return errors.Wrapf(err, "network")
+	case err = <-clientErrCh:
+		fmt.Println("client terminated")
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/roothash_messages.go]: Remove this (I noticed there is also a leftover print in the other scenario, can you remove that while you're here).

File: go/staking/api/api.go [hunk 3]
```diff
@@ -352,6 +352,10 @@ const (
 	KindStorage   ThresholdKind = 3
 
 	KindMax = KindStorage
+
+	// LogEventGeneralAdjustment is a log event value that signals adjustment
+	// of an account's general balance due to a roothash message.
+	LogEventGeneralAdjustment = "staking/general_adjustment"
```
Comment (kostko) [inline go/staking/api/api.go]: I would move this to the const block at the top as this one is really threshold-kind-related.

File: go/tendermint/apps/staking/state.go [hunk 4]
```diff
@@ -35,6 +37,12 @@ var (
 	//
 	// Value is a little endian encoding of an uint64.
 	debondingIntervalKeyFmt = keyformat.New(0x54)
+	// acceptableTransferPeersKeyFmt is the key format used for the acceptable transfer peers set.
+	//
+	// Value is a CBOR-serialized map from acceptable runtime IDs the boolean true.
```
Comment (kostko) [inline go/tendermint/apps/staking/state.go]:
```suggestion
	// Value is a CBOR-serialized map from acceptable runtime IDs to the boolean true.
```

File: go/tendermint/apps/staking/state.go [hunk 5]
```diff
@@ -35,6 +37,12 @@ var (
...
+	acceptableTransferPeersKeyFmt = keyformat.New(0x55)
```
Comment (kostko) [inline go/tendermint/apps/staking/state.go]: I assume the size of this map is expected to be small (on the order of number of runtimes)? Another option would be to have the entries be stored in the tree -- e.g., one key per entry:
```golang
keyformat.New(0x55, &signature.MapKey{})
```

File: go/worker/merge/committee/node.go [hunk 6]
```diff
@@ -464,9 +464,20 @@ func (n *Node) startMergeLocked(commitments []commitment.ComputeCommitment, resu
 		defer cancel()
 
 		var ioRoots, stateRoots []hash.Hash
+		var roothashMessages []*block.RoothashMessage
 		for _, result := range results {
 			ioRoots = append(ioRoots, result.IORoot)
 			stateRoots = append(stateRoots, result.StateRoot)
+
+			// Merge roothash messages.
+			// The rule is that at most one result can have sent roothash messages.
```
Comment (kostko) [inline go/worker/merge/committee/node.go]: Added to [list of slashing conditions](https://github.com/oasislabs/oasis-core/issues/2078) for the transaction scheduler.

File: runtime/src/transaction/context.rs [hunk 7]
```diff
@@ -72,4 +76,8 @@ impl<'a> Context<'a> {
             .expect("tags is not empty")
             .push(Tag::new(key.as_ref().to_vec(), value.as_ref().to_vec()))
     }
+
+    pub fn send_roothash_message(&mut self, roothash_message: RoothashMessage) {
```
Comment (kostko) [inline runtime/src/transaction/context.rs]: Add documentation comment.

File: go/oasis-node/cmd/debug/byzantine/compute.go (hunk 1 above)
Comment (pro-wh) [reply go/oasis-node/cmd/debug/byzantine/compute.go]: ok #2269

File: go/oasis-test-runner/scenario/e2e/roothash_messages.go (hunk 2 above)
Comment (pro-wh) [reply go/oasis-test-runner/scenario/e2e/roothash_messages.go]: if we don't want to know, then ok

File: go/staking/api/api.go (hunk 3 above)
Comment (pro-wh) [reply go/staking/api/api.go]: ok

File: go/tendermint/apps/staking/state.go (hunk 4 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/state.go]: thanks

File: go/worker/merge/committee/node.go (hunk 6 above)
Comment (pro-wh) [reply go/worker/merge/committee/node.go]: thanks

File: runtime/src/transaction/context.rs (hunk 7 above)
Comment (pro-wh) [reply runtime/src/transaction/context.rs]: ok

File: go/tendermint/apps/staking/state.go (hunk 5 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/state.go]: Right, that's the expectation. Thanks for posting another option. Let's keep that in mind so we can later support a large number of acceptable runtimes if we want to.

**Comment by pro-wh** [comment]
`cargo audit` report shows now new problems compared to master. current problems, common with master:

- memoffset 0.2.1 https://rustsec.org/advisories/RUSTSEC-2019-0011
- protobuf 1.7.4 https://rustsec.org/advisories/RUSTSEC-2019-0003
- protobuf 2.4.2 https://rustsec.org/advisories/RUSTSEC-2019-0003
- smallvec 0.6.9 https://rustsec.org/advisories/RUSTSEC-2019-0009, https://rustsec.org/advisories/RUSTSEC-2019-0012
- spin 0.5.0 https://rustsec.org/advisories/RUSTSEC-2019-0013
- rust-crypto 0.2.36 https://rustsec.org/advisories/RUSTSEC-2016-0005

**Pull request merged**